rofl-scheduler/src/manager.rs

```diff
@@ -0,0 +1,1334 @@
+    /// A list of instance identifiers that should have no running deployments.
+    pending_stop: Vec<(InstanceId, bool)>,
+    /// A map of instance updates.
+    instance_updates: BTreeMap<InstanceId, InstanceUpdates>,
+    /// A list of instance identifiers that should be accepted.
+    accept: Vec<InstanceId>,
+    /// A list of not-accepted instance identifiers and timestamps that should maybe be removed.
+    maybe_remove: Vec<(InstanceId, u64)>,
+    /// A list of instances to claim payment for.
+    claim_payment: Vec<InstanceId>,
+    /// Amounts of resources used.
+    resources_used: Resources,
+}
+
+#[derive(Default)]
+struct InstanceState {
+    /// Last deployment.
+    last_deployment: Option<Deployment>,
+    /// Last error message corresponding to deploying `last_deployment`.
+    last_error: Option<String>,
+    // Whether to ignore instance start until the given time elapses.
+    ignore_start_until: Option<Instant>,
+    /// Backoff associated with ignoring instance start.
+    ignore_start_backoff: Option<backoff::ExponentialBackoff>,
+}
+
+struct DeploymentInfo {
+    temporary_name: String,
+    manifest_hash: Hash,
+    volumes: Vec<String>,
+}
+
+/// Instance manager.
+pub struct Manager {
+    env: Environment<SchedulerApp>,
+    client: Arc<MarketClient>,
+    cfg: LocalConfig,
+
+    instances: Mutex<BTreeMap<InstanceId, InstanceState>>,
+
+    logger: slog::Logger,
+}
+
+impl Manager {
+    /// Create a new manager instance.
+    pub fn new(env: Environment<SchedulerApp>, cfg: LocalConfig) -> Self {
+        Self {
+            client: Arc::new(MarketClient::new(env.clone(), cfg.provider_address)),
+            env,
+            cfg,
+            instances: Mutex::new(BTreeMap::new()),
+            logger: get_logger("scheduler/manager"),
+        }
+    }
+
+    /// Main loop of the ROFL scheduler.
+    pub async fn run(self) {
+        let mgr = Arc::new(self);
+        let local_node_id = match mgr.env.host().identity().await {
+            Ok(local_node_id) => local_node_id,
+            Err(err) => {
+                slog::error!(mgr.logger, "failed to determine local node ID";
+                    "err" => ?err,
+                );
+                process::abort();
+            }
+        };
+        let mut last_round = 0;
+
+        loop {
+            // Wait a bit before doing another pass.
+            time::sleep(time::Duration::from_secs(mgr.cfg.processing_interval_secs)).await;
+
+            // Discover local state.
+            let mut local_state = match mgr.discover(local_node_id).await {
+                Ok(local_state) => local_state,
+                Err(err) => {
+                    slog::error!(mgr.logger, "failed to discover bundles"; "err" => ?err);
+                    continue;
+                }
+            };
+
+            // Make sure to not re-process the same round multiple times.
+            if local_state.client.round() <= last_round {
```

Comment:
Should this loop rather just watch/poll rounds? Instead of having an interval configuration and then also checking that same round is not processed multiple times?

Reply:
I don't think it makes much difference as it needs to poll other stuff anyway.

Reply:
I see. I just don't see much usefulness of `processing_interval_secs` config option, since we are limited to processing at most once per round?

Reply:
Currently yes, but the plan is for the scheduler to also expose a direct endpoint so commands can be delivered to it off-chain. This is also required for streaming logs.